### ISSUE TITLE
[2.1] Move `Module::runAction()` to `Application`

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#14855]
Currently `runAction` is implemented in `Module`.
However it is only ever called for root modules, which are `Application`s.

This won't BC in most applications but since it doesn't functionally add anything I propose doing this in 2.1

**Comment by samdark** [comment]
Makes sense to me. @yiisoft/core-developers thoughts?

**Comment by klimov-paul** [comment]
It is need to be looked carefully. The actual solution should be developed in the scope of #13922.
One thing can be said for certain: the threatment of the action result, like converting string or stream into `yii\web\Response` or integer to `yii\console\Response` should be perfromed at Application instead of Module.

**Comment by SamMousa** [comment]
@klimov-paul this is not about a functional change, just about a scope change.
The current scope for `runAction` is `Module`, but it is never called other than by an `Application`.

**Comment by klimov-paul** [comment]
> The current scope for runAction is Module, but it is never called other than by an Application.

That is unlikely to be possible, since `runAction()` includes instantiating controller and action instances, which search depends on the module. Module has the knoledge about controller path and may resolve route in his own way - such ability should not be dropped.

**Comment by SamMousa** [comment]
Actually that's `createController`.

**Comment by klimov-paul** [comment]
It is called inside `runAction()` and is just a particular implementation of route handling:
https://github.com/yiisoft/yii2/blob/master/framework/base/Module.php#L522

Module for instance may handle all incoming routes and process them in some default manner instead of throwing `InvalidRouteException` or handle route without any controller or action involved.

**Comment by SamMousa** [comment]
Yes, but it's `createController` that calls the modules' `createController`, `createAction` is never called except for the root module. 
As far as I can tell `runAction` is never called for a module that is not the root module.
All I'm proposing is that we move `runAction` to `Application`. I'm not proposing touching `createController`...

**Comment by dynasource** [comment]
rather than moving stuff away from Module.php, I would prefer the opposite. The more functionality is available in Module.php, the more can be reused in (module) extensions.

**Comment by SamMousa** [comment]
I feel there is a misconception; this specific part is not reusable at
all...

Try overriding it in a module and see what happens...

**Comment by dynasource** [comment]
> this specific part is not reusable at
> all...

please elaborate

**Comment by SamMousa** [comment]
As I mentioned in the original post, the function is never called in a
module only in an application.
